File: solutions/ocaml/032-combination-sum.ml
(*
  Problem 32: Combination Sum (LeetCode 39)
  Difficulty: Med
  Language: OCaml
*)
let combination_sum candidates target =
  let sorted = List.sort compare candidates in
  let rec go cands remain curr =
    match cands with
    | [] -> if remain = 0 then [List.rev curr] else []
    | c :: _ when c > remain ->
      if remain = 0 then [List.rev curr] else []
    | c :: rest ->
      if remain = 0 then [List.rev curr]
      else go cands (remain - c) (c :: curr) @ go rest remain curr
  in
  go sorted target []

File: solutions/scala/032-combination-sum.scala
/**
 * Problem 32: Combination Sum (LeetCode 39)
 * Difficulty: Med
 * Language: Scala
 */
def combinationSum(candidates: Array[Int], target: Int): List[List[Int]] = {
  val sorted = candidates.sorted
  def go(start: Int, remain: Int, curr: List[Int]): List[List[Int]] = {
    if (remain == 0) List(curr.reverse)
    else if (start >= sorted.length || sorted(start) > remain) Nil
    else go(start, remain - sorted(start), sorted(start) :: curr) ++
         go(start + 1, remain, curr)
  }
  go(0, target, Nil)
}
